pkg/controllers/membercluster/v1beta1/membercluster_controller.go

```diff
@@ -162,37 +171,56 @@ func (r *Reconciler) getInternalMemberCluster(ctx context.Context, name string)
-	var works placementv1beta1.WorkList
-	var clusterNS corev1.Namespace
-	// check if the namespace still exist
-	namespaceName := fmt.Sprintf(utils.NamespaceNameFormat, mc.Name)
-	if err := r.Client.Get(ctx, types.NamespacedName{Name: namespaceName}, &clusterNS); apierrors.IsNotFound(err) {
-		klog.V(2).InfoS("the member cluster namespace is successfully deleted", "memberCluster", klog.KObj(mc))
-		return runtime.Result{}, nil
-	}
+func (r *Reconciler) garbageCollectWork(ctx context.Context, mc *clusterv1beta1.MemberCluster, namespaceName string) error {
 	// list all the work object we created in the member cluster namespace
+	var works placementv1beta1.WorkList
 	listOpts := []client.ListOption{
-		client.MatchingLabels{utils.LabelFleetObj: utils.LabelFleetObjValue},
 		client.InNamespace(namespaceName),
 	}
 	if err := r.Client.List(ctx, &works, listOpts...); err != nil {
 		klog.ErrorS(err, "failed to list all the work object", "memberCluster", klog.KObj(mc))
-		return runtime.Result{}, client.IgnoreNotFound(err)
+		return client.IgnoreNotFound(err)
 	}
+	// remove all the finalizers on the work objects in parallel
+	errs, cctx := errgroup.WithContext(ctx)
 	for _, work := range works.Items {
 		staleWork := work.DeepCopy()
-		staleWork.SetFinalizers(nil)
-		if updateErr := r.Update(ctx, staleWork, &client.UpdateOptions{}); updateErr != nil {
-			klog.ErrorS(updateErr, "failed to remove the finalizer from the work",
-				"memberCluster", klog.KObj(mc), "work", klog.KObj(staleWork))
-			return runtime.Result{}, updateErr
-		}
+		errs.Go(func() error {
+			staleWork.SetFinalizers(nil)
+			if updateErr := r.Update(cctx, staleWork, &client.UpdateOptions{}); updateErr != nil {
+				klog.ErrorS(updateErr, "failed to remove the finalizer from the work",
+					"memberCluster", klog.KObj(mc), "work", klog.KObj(staleWork))
+				return updateErr
+			}
+			return nil
+		})
 	}
-	klog.V(2).InfoS("successfully removed all the work finalizers in the cluster namespace",
+	klog.V(2).InfoS("try to remove all the work finalizers in the cluster namespace",
 		"memberCluster", klog.KObj(mc), "number of work", len(works.Items))
+	return errs.Wait()
+}
+
+// garbageCollect is used to garbage collect all the resources in the cluster namespace associated with the member cluster.
+func (r *Reconciler) garbageCollect(ctx context.Context, mc *clusterv1beta1.MemberCluster) error {
+	// check if the namespace still exist
+	var clusterNS corev1.Namespace
+	namespaceName := fmt.Sprintf(utils.NamespaceNameFormat, mc.Name)
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: namespaceName}, &clusterNS); apierrors.IsNotFound(err) {
+		klog.V(2).InfoS("the member cluster namespace is successfully deleted", "memberCluster", klog.KObj(mc))
+		return nil
+	}
+	if err := r.garbageCollectWork(ctx, mc, namespaceName); err != nil {
+		return err
+	}
+	// delete the namespace in the foreground so that we are sure it is deleted before we mark the member cluster as left
```

Comment:
kss delete won't be blocked because some resources have owner not deleted (owner deleted will trigger gc but not the other way around). Here I remove the finalizer from MC after the NS is deleted so the MC won't be removed until the NS is deleted.

Reply:
hmm, ```Here I remove the finalizer from MC after the NS is deleted so the MC won't be removed until the NS is deleted.```

but you did not wait here. line 217 is not a blocking call to make sure the ns is deleted before return.